hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -190,55 +190,56 @@ object InferPType {
         as.foreach(infer(_))
 
-        infer(body, env.bindIterable(names.zip(as.map(_.pType2.asInstanceOf[PArray].elementType))))
+        infer(body, env.bindIterable(names.zip(as.map(_.pType2.asInstanceOf[PStreamable].elementType))))
         coerce[PStreamable](as.head.pType2).copyStreamable(body.pType2, as.forall(_.pType2.required))
       case ArrayFilter(a, name, cond) =>
         infer(a)
         a.pType2
       case ArrayFlatMap(a, name, body) =>
         infer(a)
-        infer(body, env.bind(name, a.pType2.asInstanceOf[PArray].elementType))
+        infer(body, env.bind(name, a.pType2.asInstanceOf[PStreamable].elementType))
 
         // Whether an array must return depends on a, but element requiredeness depends on body (null a elements elided)
         coerce[PStreamable](a.pType2).copyStreamable(coerce[PIterable](body.pType2).elementType, a.pType2.required)
       case ArrayFold(a, zero, accumName, valueName, body) =>
         infer(zero)
 
         infer(a)
-        infer(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PArray].elementType))
+        infer(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PStreamable].elementType))
         assert(body.pType2 isOfType zero.pType2)
 
         zero.pType2.setRequired(body.pType2.required)
       case ArrayFold2(a, acc, valueName, seq, res) =>
         infer(a)
         acc.foreach { case (_, accIR) => infer(accIR) }
         val resEnv = env.bind(acc.map { case (name, accIR) => (name, accIR.pType2) }: _*)
-        val seqEnv = resEnv.bind(valueName -> a.pType2.asInstanceOf[PArray].elementType)
+        val seqEnv = resEnv.bind(valueName -> a.pType2.asInstanceOf[PStreamable].elementType)
         seq.foreach(infer(_, seqEnv))
         infer(res, resEnv)
         res.pType2.setRequired(res.pType2.required && a.pType2.required)
       case ArrayScan(a, zero, accumName, valueName, body) =>
         infer(zero)
 
         infer(a)
-        infer(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PArray].elementType))
+        infer(body, env.bind(accumName -> zero.pType2, valueName -> a.pType2.asInstanceOf[PStreamable].elementType))
         assert(body.pType2 isOfType zero.pType2)
 
         val elementPType = zero.pType2.setRequired(body.pType2.required && zero.pType2.required)
         coerce[PStreamable](a.pType2).copyStreamable(elementPType, a.pType2.required)
       case ArrayLeftJoinDistinct(lIR, rIR, lName, rName, compare, join) =>
         infer(lIR)
         infer(rIR)
-        val e = env.bind(lName -> lIR.pType2.asInstanceOf[PArray].elementType, rName -> rIR.pType2.asInstanceOf[PArray].elementType)
+        val e = env.bind(lName -> lIR.pType2.asInstanceOf[PStreamable].elementType, rName -> rIR.pType2.asInstanceOf[PStreamable].elementType)
 
         infer(compare, e)
         infer(join, e)
 
+        // wrong
         PArray(join.pType2, lIR.pType2.required)
```

Comment:
Should this be IR.pType2.asInstanceOf[PStreamable].copyStreamable(join.pType2, lIR.pType2.required)? Or, in the case we decide to go with PStream as you recommended (I'm having some tests failures), the same with asInstanceOf[PStream].

Similarly, InferType has a broken implementation:

```scala
case ArrayLeftJoinDistinct(left, right, l, r, compare, join) =>
        coerce[TStreamable](left.typ).copyStreamable(join.typ)
        TArray(join.typ)
```

should be either TArray(join.typ), or the copyStreamable, which is currently discarded.

Reply:
yes, just delete the second line, I think.

Reply:
the ptype rule should probably be `PStream(join.pType2, l.pType2.required && r.pType2.required)`

Reply:
Yeah, I'm doing this everywhere that was casting to PStream. Some test failures. I think it would be better with a .copy instead, to handle a future case with a non-canonical stream?

Reply:
either one, up to you. I don't think copy handles future-proofing any better  though.

Reply:
Out of curiosity, why do you think that? Wouldn't this allow us to instantiate based on the inferred ptype (using the inferred ptype's constructor) rather than a canonical ptype?